feat: fast txs request implementation

mralj opened the pull request:
Adds fast tx request per this spec: https://github.com/AztecProtocol/aztec-packages/issues/15149

ludamad commented on yarn-project/foundation/src/promise/utils.ts:
```diff
@@ -27,3 +29,25 @@ export function promiseWithResolvers<T>(): PromiseWithResolvers<T> {
+ * @param timeout - The maximum time in milliseconds to wait before rejecting (default: 5000ms)
+ * @param dateProvider - An optional DateProvider instance for getting the current time (default: new DateProvider())
+ */
+export function waitFor(pred: () => boolean, interval = 10, timeout = 5_000, dateProvider = new DateProvider()) {
```
Seems redundant with `retryUntil`. Also I don't see how this is a promise utility

just-mitch commented:
Fix A-114

ludamad commented:
Curious, why no test cache? (as that will e.g. run bb prover tests everytime)

mralj commented:
> Curious, why no test cache? (as that will e.g. run bb prover tests everytime)

Sorry @ludamad this is stupid mistake on my part, removed label

PhilWindle commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.test.ts:
```diff
@@ -0,0 +1,1790 @@
+
+      // Second round should be [25, 0-6] - because we wrap around to make sure we always request TX_BATCH_SIZE,
+      // [0-7], [8-15]
+      expect(peer1Requests[1].indices).toEqual([
```
Slightly confused why this isn't `[16-24]`?

mralj replied:
// With txCount=25 and TX_BATCH_SIZE=8, we get 4 chunks after wrap-around padding:
      // Chunk 0: [0-7], Chunk 1: [8-15], Chunk 2: [16-23], Chunk 3: [24, 0-6]
      // (Chunk 3 wraps around to ensure every batch has exactly TX_BATCH_SIZE items)
      //
      // Workers share a round-robin index that advances globally across all workers.
      //
      // Round 1: Workers take chunks 0, 1, 2 respectively
      expect(peer1Requests[0].indices).toEqual(Array.from({ length: TX_BATCH_SIZE }, (_, i) => i));
      expect(peer2Requests[0].indices).toEqual(Array.from({ length: TX_BATCH_SIZE }, (_, i) => i + TX_BATCH_SIZE));
      expect(peer3Requests[0].indices).toEqual(Array.from({ length: TX_BATCH_SIZE }, (_, i) => i + 2 * TX_BATCH_SIZE));

      // Round 2: The shared round-robin index continues from where it left off (index 3).
      // - Peer1 takes chunk 3: [24, 0-6], then index wraps to 0
      // - Peer2 takes chunk 0: [0-7], index advances to 1
      // - Peer3 takes chunk 1: [8-15], index advances to 2
      // This cycling is intentional. It re-requests batches from different peers to maximize
      // chances of fetching missing transactions.
      expect(peer1Requests[1].indices).toEqual([
        ...Array.from({ length: TX_BATCH_SIZE - 1 }, (_, i) => i),
        txCount - 1,
      ]);
      expect(peer2Requests[1].indices).toEqual(Array.from({ length: TX_BATCH_SIZE }, (_, i) => i));
      expect(peer3Requests[1].indices).toEqual(Array.from({ length: TX_BATCH_SIZE }, (_, i) => i + TX_BATCH_SIZE));
    });

I have pushed the commit, giving a better explanation of what's happening here. Here is C/P.

PhilWindle commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/missing_txs.ts:
```diff
@@ -0,0 +1,210 @@
+ * This could be better optimized but given expected count of missing txs (N < 100)
+ * At the moment there is no need for it. And benefit is that we have everything in single store
+ * */
+export class MissingTxMetadataCollection extends Map<string, MissingTxMetadata> implements ITxMetadataCollection {
```
With this class and the one above. We do a lot of `TxHash.toString()`. Is it better to keep `MissingTxMetadata`'s member as a `TxHash` or would it be more optimal to store a `string`?

PhilWindle commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+import { MissingTxMetadata, MissingTxMetadataCollection, TX_BATCH_SIZE } from './missing_txs.js';
+import { type IPeerCollection, PeerCollection, RATE_LIMIT_EXCEEDED_PEER_CACHE_TTL } from './peer_collection.js';
+
+const SMART_PEERS_TO_QUERY_IN_PARALLEL = 10;
```
We will want to make these defaults configuration options I think.

PhilWindle commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1431,6 +1442,37 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+        const _severity = await this.handleDoubleSpendFailure(tx, txBlockNumber);
+      }
+
+      // TODO: Penalize Peer?
```
I think we definitely want to penalise the peer here.

spalladino replied:
Do we? It was this node who asked for that tx, why should we penalize the peer who replied to the request. I think we should only penalize if the resulting tx fails some very basic checks, like hash doesn't match or proof is invalid.

PhilWindle commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1431,6 +1442,37 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+  @trackSpan('Libp2pService.validateFastTx', tx => ({
+    [Attributes.TX_HASH]: tx.getTxHash().toString(),
+  }))
+  private async validateFastTx(tx: Tx, _peerId: PeerId): Promise<boolean> {
```
I'm confused as to all of the `validateXYZTx` functions we have now and what they are used for.

Firstly, this is using the validator returned from `createMessageValidators`. I think this is wrong. That validator is for gossiped transactions and includes checks that I don't think we should eb doing here.

However, looking through the code more broadly and I can see a number of issues. So I am thinking that it needs tackling as part of my wider effort around the mempool.

spalladino replied:
Agree. This can also lead to a loop where we request a tx, get it back, reject it because we considered it invalid, requested it again, etc.

PhilWindle commented on yarn-project/p2p/src/services/tx_collection/fast_tx_collection.ts:
```diff
@@ -261,16 +263,34 @@ export class FastTxCollection {
+
+            return await BatchTxRequester.collectAllTxs(batchRequester.run());
+          } else if (request.type === 'block') {
+            const txs = await this.p2pService.reqResp.sendBatchRequest<ReqRespSubProtocol.TX>(
```
Is this the old req/resp system?

mralj replied:
Yes.

spalladino commented on yarn-project/p2p/src/services/libp2p/libp2p_service.ts:
```diff
@@ -1431,6 +1442,37 @@ export class LibP2PService<T extends P2PClientType = P2PClientType.Full> extends
+    const currentBlockNumber = await this.archiver.getBlockNumber();
+
+    // We accept transactions if they are not expired by the next slot (checked based on the IncludeByTimestamp field)
+    const { ts: nextSlotTimestamp } = this.epochCache.getEpochAndSlotInNextL1Slot();
```
This may break provers requesting txs that have already been mined in an older block in the epoch, when they were still valid. Same applies to the `epoch_prune_watcher`, which tries reexecuting a past epoch.

We probably need to include some sort of context for this validate function, so we know whether we are requesting txs for inclusion or reexecution.

spalladino commented on yarn-project/p2p/src/services/reqresp/protocols/block_txs/block_txs_reqresp.ts:
```diff
@@ -9,12 +10,52 @@ import { BitVector } from './bitvector.js';
+
+    const missingHashesSet = new Set(missingTxHashes.map(t => t.toString()));
+
+    // We cannot request txs that are not part of the block proposal
+    if (!missingHashesSet.isSubsetOf(new Set(blockProposal.txHashes.map(t => t.toString())))) {
+      return undefined;
+    }
```
Shouldn't we intersect the sets, rather than just returning undefined? Or is this handled by the caller?

mralj replied:
Handled by a caller - the idea is that if something weird or bad happens, and for whatever weird reason we ask for txs that aren't part of the block, we return `undefined` so we don't send this request to the peer.

spalladino commented on yarn-project/p2p/src/services/reqresp/reqresp.ts:
```diff
@@ -409,6 +416,7 @@ export class ReqResp implements ReqRespInterface {
       );
       return resp;
     } catch (e: any) {
+      this.logger.warn(`SUBPROTOCOL: ${subProtocol}\n`, e);
```
Leftover from debugging?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/README.md:
```diff
@@ -0,0 +1,303 @@
+| `smartParallelWorkerCount` | 10 | Max concurrent requests to smart peers |
+| `dumbParallelWorkerCount` | 10 | Max concurrent requests to dumb peers |
+| `TX_BATCH_SIZE` | 8 | Max transactions per request |
+| `BAD_PEER_THRESHOLD` | 3 | Failures before marking peer as bad |
```
I understand what's a pinned, dumb, and smart peer. But what is a "bad" peer?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/README.md:
```diff
@@ -0,0 +1,303 @@
+
+### Peer States
+
+- **Bad Peer**: After `BAD_PEER_THRESHOLD` consecutive failures, peer is excluded from queries
```
Ah, I see. Still, what's a failure? An invalid response? An empty one?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/README.md:
```diff
@@ -0,0 +1,303 @@
+                    │                                 │
+                    ▼                                 ▼
+┌───────────────────────────────────┐   ┌─────────────────────────────────────┐
+│        FastTxCollection           │   │         SlowTxCollection            │
```
IIRC fast tx collection is sometimes also used by the prover node when it needs to start proving and doesn't have all txs.

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/missing_txs.ts:
```diff
@@ -0,0 +1,210 @@
+    );
+  }
+
+  public getSortedByRequestedCountThenByInFlightCountAsc(txs: string[]): MissingTxMetadata[] {
```
I think the name doesn't match what the function does. I'd've expected this to sort by requestCount, and use inFlightCount as tie breaker.

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/missing_txs.ts:
```diff
@@ -0,0 +1,210 @@
+    // If something went wrong and we don't have txMeta for this hash
+    // We should not request it, so here we "pretend" that it was fetched
+    return this.get(txHash.toString())?.fetched ?? true;
+  }
```
I'd remove/rename this method. We have both `isFetched` and `alreadyFetched`, which do the same but handle the error case differently (without any warning), and to make it worse, `isFetched` may return true for txs not in the result of `getFetchedTxs`. It's pretty confusing.

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/peer_collection.ts:
```diff
@@ -0,0 +1,129 @@
+
+import type { PeerId } from '@libp2p/interface';
+
+export const BAD_PEER_THRESHOLD = 3;
+export const RATE_LIMIT_EXCEEDED_PEER_CACHE_TTL = 1000; // 1s
+
+export interface IPeerCollection {
+  getAllPeers(): Set<string>;
+  getSmartPeers(): Set<string>;
+  markPeerSmart(peerId: PeerId): void;
+  getSmartPeersToQuery(): Array<string>;
+  getDumbPeersToQuery(): Array<string>;
+  thereAreSomeDumbRatelimitExceededPeers(): boolean;
+  markPeerAsBad(peerId: PeerId): void;
+  unMarkPeerAsBad(peerId: PeerId): void;
+  getBadPeers(): Set<string>;
+  markPeerInFlight(peerId: PeerId): void;
+  unMarkPeerInFlight(peerId: PeerId): void;
+  markPeerRateLimitExceeded(peerId: PeerId): void;
+  getRateLimitExceededPeers(): Set<string>;
+}
```
If we don't have more than one implementation for this interface, let's just kill it and use the class directly. It makes "go-to-definition" easier.

mralj replied:
I'd rather keep it.
The other one is in the tests `export class TestPeerCollection implements IPeerCollection. 
I use it to help with tracking of "some stuff" for easier logging and testing, for example:
```TS
  markPeerInFlight(peerId: any): void {
    this.peersMarkedInFlight.push(peerId.toString());
    return this.inner.markPeerInFlight(peerId);
  }
```

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/peer_collection.ts:
```diff
@@ -0,0 +1,129 @@
+
+  public markPeerAsBad(peerId: PeerId) {
+    this.badPeers.set(peerId.toString(), (this.badPeers.get(peerId.toString()) ?? 0) + 1);
+  }
```
Can we rename this method? I thought it marked the peer as bad, but it actually increased the "badness" counter, which may not lead to the peer being bad just yet.

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+   * */
+  private async handleSuccessResponseFromPeer(peerId: PeerId, response: BlockTxsResponse) {
+    // If we have received successful response from the peer, they have "redeemed" themselves and not considered bad anymore
+    this.peers.unMarkPeerAsBad(peerId);
```
Shouldn't we unmark the peer as bad only if the txs they sent pass a validation?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+      const wrapAroundCount = TX_BATCH_SIZE - remainder;
+      const wrappedHashes = [...missingHashes, ...missingHashes.slice(0, wrapAroundCount)];
+      return chunk<string>(wrappedHashes, TX_BATCH_SIZE);
+    };
```
Feels like this could be extracted to an array function in foundation

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+        return undefined;
+      }
+
+      if (chunks[idx] === undefined) {
```
Is this possible?

mralj replied:
Nope, removed

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+          } else {
+            // There are still some dumb peers to query but they have been rate limited
+            // Sleep until they get unblocked
+            await sleep(RATE_LIMIT_EXCEEDED_PEER_CACHE_TTL);
```
We should be sleeping the min amount of time until the first peer gets unblocked, not the full TTL

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+        if (weRanOutOfPeersToQuery) {
+          if (!this.peers.thereAreSomeDumbRatelimitExceededPeers()) {
+            this.logger.debug(`Worker loop dumb: No more peers to query`);
+            break;
```
Can we get more dumb peers to show up somehow? Is it feasible that peer discovery kicks in during fast collection and we get a new peer to play with? Should we sleep instead of break?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+              `Worker loop dumb: Could not create next batch request ${FAILED_MAKE_REQUEST_THRESHOLD} times, exiting`,
+            );
+            break;
+          }
+
+          this.logger.debug(`Worker loop dumb: Could not create next batch request`);
+          continue;
+        }
```
Why can a `makeRequest` actually fail? Does it make sense to track this threshold, or should we just retry?

mralj replied:
This was pretty stupid of me. I have no idea what went through my head when I wrote this.
Anyway removed this part completely. 

And to answer your question, it's not that `makeRequest` actually "fails" - it is that there are no more txs to request  - there are 2 scenarios:
1. We got them all already
2. This one - I have no idea why/how it culd happen in practice, but in theory I guess it could if something went really bad somewhere:

```TS
      const blockRequest = BlockTxsRequest.fromBlockProposalAndMissingTxs(
        this.blockProposal,
        txs,
        includeFullHashesInRequestNotJustIndices,
      );
```

This is unable to produce blockRequest because the hashes we provided are not part of the `block.Proposal`.

But I have no idea why I thought that retrying here would help ... ever.

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+    };
+
+    const workers = Array.from(
+      { length: Math.min(this.opts.smartParallelWorkerCount, this.peers.getAllPeers().size) },
```
Same question as before, can we get new peers while running this?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+        });
+      }
+    } catch (err: any) {
+      this.logger.error(`Smart worker ${workerIndex} encountered an error: ${err}`);
```
Should we handle the timeout error acquiring semaphore differently, given it's not actually an error?

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+            // These might be either smart peers that will get unblocked after _some time_
+            const thereAreSomeRateLimitedPeers = this.peers.getRateLimitExceededPeers().size > 0;
+            if (thereAreSomeRateLimitedPeers) {
+              await sleep(RATE_LIMIT_EXCEEDED_PEER_CACHE_TTL);
```
Same, should not sleep the full TTL

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+
+          // Otherwise there are still some dumb peers that could become smart
+          // We wait on this
+          await executeTimeout((_: AbortSignal) => this.smartRequesterSemaphore.acquire(), this.timeoutMs);
```
Doesn't this lead to this smart worker loop acquiring the semaphore more than once?

mralj replied:
This is a correct observation. I have updated the comment to better explain why this is needed, here C/P of newly added code comment:

```ts
          // We end up here when all known smart peers became temporarily unavailable via combination of
          // (bad, in-flight, or rate-limited) or in some weird scenario all current smart peers turn bad which is permanent
          // but dumb peers still exist that could become smart.
          //
          // When a dumb peer responds with valid txIndices, it gets
          // promoted to smart and releases the semaphore, waking this worker.

```

But while double-checking this, I have acutally found bug, before we had this release:
```ts
// Unblock smart workers
if (this.peers.getSmartPeers().size <= this.opts.smartParallelWorkerCount) 
```

Whereas it should be
```ts
// Unblock smart workers
    if (this.peers.getSmartPeersToQuery().length <= this.opts.smartParallelWorkerCount) 
```

Key difference being `getSmartPeers` vs `getSmartPeersToQuery`

spalladino commented on yarn-project/p2p/src/services/reqresp/batch-tx-requester/batch_tx_requester.ts:
```diff
@@ -0,0 +1,670 @@
+      const request = BlockTxsRequest.fromBlockProposalAndMissingTxs(this.blockProposal, txs);
+      if (!request) {
+        return;
+      }
```
We should only flag the txs as in flight and requested if the request was created successfully

spalladino reviewed (commented):
Left a bunch of comments, but I think the only critical one is the one that Phil pointed out on running overzealous validators on the returned txs.

That said, I'm worried about the increasing complexity of all this. We have the tx collector, which has a loop for slowly retrieving txs, plus an action for fast retrieving them from nodes and reqresp. These two communicate with each other via the tx sink to make sure there are no duplicate efforts. But then the fast collector may jump into the batch tx requester, which starts its own three loops, with no communication back to the sink; or may jump into the `reqresp.sendBatchRequest` for fetching txs outside the context of a block, which leads to reqresp own batching and looping.

I'm not sure how to solve the above, but I'm concerned about how we'll troubleshoot this with some many moving pieces. FWIW I'm fine merging this after we fix the most pressing issues, and adding a config variable to toggle between this protocol and the legacy one should we need to fall back to that one.

mralj commented:
@PhilWindle  @spalladino  I have addressed most of the comments; a couple of things still need finishing touches. I hve openned following issues to track them and assigned them all to @deffrian ([list](https://linear.app/aztec-labs/project/alpha-network-performance-70eb00d71102/issues))

mralj committed:
feat: fast txs request implementation

Adds fast tx request per this spec: https://github.com/AztecProtocol/aztec-packages/issues/15149
